File: constructionRaceway.cpp
#include <bits/stdc++.h>
using namespace std;

int check[26][26];
int dx[] = {-1, 0, 1, 0};
int dy[] = {0, 1, 0, -1};
int solution(vector<vector<int>> a) {
    int n = a.size();
    queue<tuple<int, int, int, int>> q;
    q.push({0, 0, -1, 0});
    check[0][0] = 1;
    while (!q.empty()) {
        auto [x, y, d, cost] = q.front();
        q.pop();
        for (int k = 0; k < 4; k++) {
            int ncost = cost + 100;
            if (d != -1 && d != k) ncost += 500;
            int nx = x + dx[k];
            int ny = y + dy[k];
            if (0 <= nx && nx < n && 0 <= ny && ny < n && a[nx][ny] == 0) {
                if (check[nx][ny] == 0 || check[nx][ny] >= ncost) {
                    check[nx][ny] = ncost;
                    q.push({nx, ny, k, ncost});
                }
            }
        }
    }
    return check[n - 1][n - 1];
}

File: constructionRaceway.js
function solution(a) {
    let n = a.length;
    let check = [];
    let dx = [-1, 0, 1, 0];
    let dy = [0, 1, 0, -1];
    for (let i = 0; i < 26; i++) {
        check[i] = [];
        for (let j = 0; j < 26; j++) {
            check[i][j] = 2000000000;
        }
    }
    let q = [[0, 0, -1, 0]];
    check[0][0] = 0;
    while (q.length !== 0) {
        let [x, y, dir, cost] = q[0];
        q.shift();
        for (let k = 0; k < 4; k++) {
            let nx = x + dx[k];
            let ny = y + dy[k];
            let ncost = cost + 100;
            if (dir !== -1 && dir !== k) {
                ncost += 500;
            }
            if (0 <= nx && nx < n && 0 <= ny && ny < n && a[nx][ny] === 0) {
                if (check[nx][ny] >= ncost) {
                    check[nx][ny] = ncost;
                    q.push([nx, ny, k, ncost]);
                }
            }
        }
    }
    return check[n - 1][n - 1];
}
